// observable-webworker/esm5/lib/from-worker-pool.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Observable, of, Subject, zip } from 'rxjs';
import { finalize, map, mergeAll, tap } from 'rxjs/operators';
import { fromWorker } from './from-worker';
/**
 * @record
 */
function LazyWorker() { }
if (false) {
    /** @type {?} */
    LazyWorker.prototype.factory;
    /** @type {?} */
    LazyWorker.prototype.terminate;
    /** @type {?} */
    LazyWorker.prototype.processing;
    /** @type {?} */
    LazyWorker.prototype.started;
    /** @type {?} */
    LazyWorker.prototype.index;
}
/**
 * @record
 * @template I, O
 */
export function WorkerPoolOptions() { }
if (false) {
    /** @type {?|undefined} */
    WorkerPoolOptions.prototype.workerCount;
    /** @type {?|undefined} */
    WorkerPoolOptions.prototype.flattenOperator;
    /** @type {?|undefined} */
    WorkerPoolOptions.prototype.selectTransferables;
}
/**
 * @template I, O
 * @param {?} workerConstructor
 * @param {?} workUnitIterator
 * @param {?=} options
 * @return {?}
 */
export function fromWorkerPool(workerConstructor, workUnitIterator, options) {
    var _a = options || {}, 
    // tslint:disable-next-line:no-unnecessary-initializer
    _b = _a.selectTransferables, 
    // tslint:disable-next-line:no-unnecessary-initializer
    selectTransferables = _b === void 0 ? undefined : _b, _c = _a.workerCount, workerCount = _c === void 0 ? navigator.hardwareConcurrency - 1 : _c, _d = _a.flattenOperator, flattenOperator = _d === void 0 ? mergeAll() : _d;
    return new Observable((/**
     * @param {?} resultObserver
     * @return {?}
     */
    function (resultObserver) {
        /** @type {?} */
        var idleWorker$$ = new Subject();
        /** @type {?} */
        var completed = 0;
        /** @type {?} */
        var sent = 0;
        /** @type {?} */
        var finished = false;
        /** @type {?} */
        var lazyWorkers = Array.from({ length: workerCount }).map((/**
         * @param {?} _
         * @param {?} index
         * @return {?}
         */
        function (_, index) {
            return {
                _cachedWorker: null,
                factory: /**
                 * @return {?}
                 */
                function () {
                    if (!this._cachedWorker) {
                        this._cachedWorker = workerConstructor(index);
                        this.started = true;
                    }
                    return this._cachedWorker;
                },
                terminate: /**
                 * @return {?}
                 */
                function () {
                    if (this.started && !this.processing) {
                        this._cachedWorker.terminate();
                    }
                },
                processing: false,
                started: false,
                index: index,
            };
        }));
        /** @type {?} */
        var processor$ = zip(idleWorker$$, workUnitIterator).pipe(tap((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 1), worker = _b[0];
            sent++;
            worker.processing = true;
        })), finalize((/**
         * @return {?}
         */
        function () {
            idleWorker$$.complete();
            finished = true;
            lazyWorkers.forEach((/**
             * @param {?} worker
             * @return {?}
             */
            function (worker) { return worker.terminate(); }));
        })), map((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var _b = tslib_1.__read(_a, 2), worker = _b[0], unitWork = _b[1];
            return fromWorker((/**
             * @return {?}
             */
            function () { return worker.factory(); }), of(unitWork), selectTransferables, {
                terminateOnComplete: false,
            }).pipe(finalize((/**
             * @return {?}
             */
            function () {
                completed++;
                worker.processing = false;
                if (!finished) {
                    idleWorker$$.next(worker);
                }
                else {
                    worker.terminate();
                }
                if (finished && completed === sent) {
                    resultObserver.complete();
                }
            })));
        })), flattenOperator);
        /** @type {?} */
        var sub = processor$.subscribe(resultObserver);
        lazyWorkers.forEach((/**
         * @param {?} w
         * @return {?}
         */
        function (w) { return idleWorker$$.next(w); }));
        return (/**
         * @return {?}
         */
        function () { return sub.unsubscribe(); });
    }));
}
