// observable-webworker/esm5/lib/observable-worker.types.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * \@internal
 * @record
 * @template P
 */
export function GenericWorkerMessage() { }
if (false) {
    /** @type {?} */
    GenericWorkerMessage.prototype.payload;
    /** @type {?|undefined} */
    GenericWorkerMessage.prototype.transferables;
}
/**
 * \@internal
 * @record
 * @template T
 */
export function WorkerMessageNotification() { }
if (false) {
    /** @type {?} */
    WorkerMessageNotification.prototype.data;
}
/**
 * @record
 * @template I, O
 */
export function DoWorkUnit() { }
if (false) {
    /**
     * @param {?} input
     * @return {?}
     */
    DoWorkUnit.prototype.workUnit = function (input) { };
    /**
     * @param {?} output
     * @return {?}
     */
    DoWorkUnit.prototype.selectTransferables = function (output) { };
}
/**
 * @record
 * @template I, O
 */
export function DoWork() { }
if (false) {
    /**
     * @param {?} input$
     * @return {?}
     */
    DoWork.prototype.work = function (input$) { };
    /**
     * @param {?} output
     * @return {?}
     */
    DoWork.prototype.selectTransferables = function (output) { };
}
/**
 * @record
 * @template I, O
 */
export function DoTransferableWork() { }
if (false) {
    /**
     * @param {?} output
     * @return {?}
     */
    DoTransferableWork.prototype.selectTransferables = function (output) { };
}
/**
 * @record
 * @template I, O
 */
export function DoTransferableWorkUnit() { }
if (false) {
    /**
     * @param {?} output
     * @return {?}
     */
    DoTransferableWorkUnit.prototype.selectTransferables = function (output) { };
}
