// observable-webworker/esm2015/lib/from-worker-pool.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Observable, of, Subject, zip } from 'rxjs';
import { finalize, map, mergeAll, tap } from 'rxjs/operators';
import { fromWorker } from './from-worker';
/**
 * @record
 */
function LazyWorker() { }
if (false) {
    /** @type {?} */
    LazyWorker.prototype.factory;
    /** @type {?} */
    LazyWorker.prototype.terminate;
    /** @type {?} */
    LazyWorker.prototype.processing;
    /** @type {?} */
    LazyWorker.prototype.started;
    /** @type {?} */
    LazyWorker.prototype.index;
}
/**
 * @record
 * @template I, O
 */
export function WorkerPoolOptions() { }
if (false) {
    /** @type {?|undefined} */
    WorkerPoolOptions.prototype.workerCount;
    /** @type {?|undefined} */
    WorkerPoolOptions.prototype.flattenOperator;
    /** @type {?|undefined} */
    WorkerPoolOptions.prototype.selectTransferables;
}
/**
 * @template I, O
 * @param {?} workerConstructor
 * @param {?} workUnitIterator
 * @param {?=} options
 * @return {?}
 */
export function fromWorkerPool(workerConstructor, workUnitIterator, options) {
    const { 
    // tslint:disable-next-line:no-unnecessary-initializer
    selectTransferables = undefined, workerCount = navigator.hardwareConcurrency - 1, flattenOperator = mergeAll(), } = options || {};
    return new Observable((/**
     * @param {?} resultObserver
     * @return {?}
     */
    resultObserver => {
        /** @type {?} */
        const idleWorker$$ = new Subject();
        /** @type {?} */
        let completed = 0;
        /** @type {?} */
        let sent = 0;
        /** @type {?} */
        let finished = false;
        /** @type {?} */
        const lazyWorkers = Array.from({ length: workerCount }).map((/**
         * @param {?} _
         * @param {?} index
         * @return {?}
         */
        (_, index) => {
            return {
                _cachedWorker: null,
                /**
                 * @return {?}
                 */
                factory() {
                    if (!this._cachedWorker) {
                        this._cachedWorker = workerConstructor(index);
                        this.started = true;
                    }
                    return this._cachedWorker;
                },
                /**
                 * @return {?}
                 */
                terminate() {
                    if (this.started && !this.processing) {
                        this._cachedWorker.terminate();
                    }
                },
                processing: false,
                started: false,
                index,
            };
        }));
        /** @type {?} */
        const processor$ = zip(idleWorker$$, workUnitIterator).pipe(tap((/**
         * @param {?} __0
         * @return {?}
         */
        ([worker]) => {
            sent++;
            worker.processing = true;
        })), finalize((/**
         * @return {?}
         */
        () => {
            idleWorker$$.complete();
            finished = true;
            lazyWorkers.forEach((/**
             * @param {?} worker
             * @return {?}
             */
            worker => worker.terminate()));
        })), map((/**
         * @param {?} __0
         * @return {?}
         */
        ([worker, unitWork]) => {
            return fromWorker((/**
             * @return {?}
             */
            () => worker.factory()), of(unitWork), selectTransferables, {
                terminateOnComplete: false,
            }).pipe(finalize((/**
             * @return {?}
             */
            () => {
                completed++;
                worker.processing = false;
                if (!finished) {
                    idleWorker$$.next(worker);
                }
                else {
                    worker.terminate();
                }
                if (finished && completed === sent) {
                    resultObserver.complete();
                }
            })));
        })), flattenOperator);
        /** @type {?} */
        const sub = processor$.subscribe(resultObserver);
        lazyWorkers.forEach((/**
         * @param {?} w
         * @return {?}
         */
        w => idleWorker$$.next(w)));
        return (/**
         * @return {?}
         */
        () => sub.unsubscribe());
    }));
}
